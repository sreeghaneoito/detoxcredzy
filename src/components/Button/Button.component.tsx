import React, {useRef} from 'react';
import {
  ActivityIndicator,
  Animated,
  TouchableOpacity,
  View,
} from 'react-native';
import {useTheme} from '../../hooks/useTheme';
import {style} from './Button.style';
import {ButtonProps} from './../../interfaces/component.interface';

/**
 * Common Button
 * @param type: Button type. it should be primary, ghost, or danger
 * @param label: Button name
 * @param Customstyle: Custom style for button. it should be a style object
 * @param triggerclick: Function that trigger on button click
 * @param disabled: Is the button is disabled or not
 * @param testid: Test id for automation
 * @param loading: Is the button is loading or not
 * @returns JSX Component
 */
const Button = ({
  type,
  label,
  Customstyle,
  triggerclick,
  disabled,
  testid,
  loading,
}: ButtonProps) => {
  const debounceRef = useRef<any>();

  // Button Click with rebounce
  const buttonClick = () => {
    if (!debounceRef.current) {
      triggerclick();
      debounceRef.current = setInterval(() => {
        clearInterval(debounceRef.current);
        debounceRef.current = null;
      }, 800);
    }
  };

  const {theme} = useTheme();

  try {
    return (
      <TouchableOpacity
        activeOpacity={loading ? 1 : 0.2}
        testID={`${testid}button`}
        style={[
          style.btn,
          type === 'danger' && style.btndanger,
          type === 'primary' && style.btnPrimary,
          type === 'ghost' && theme.ghostBtn,
          Customstyle,
          disabled && style.disabled,
        ]}
        onPress={!disabled ? buttonClick : () => {}}>
        {loading ? (
          <Animated.View>
            <ActivityIndicator color={'#fff'} />
          </Animated.View>
        ) : (
          <Animated.Text
            style={[
              style.btnText,
              style.fillbtnText,
              type === 'ghost' && theme.onSurface,
            ]}>
            {label}
          </Animated.Text>
        )}
      </TouchableOpacity>
    );
  } catch (error) {
    console.log('Button Crash =>', error);
    return <View style={style.btn} />;
  }
};

export default Button;
