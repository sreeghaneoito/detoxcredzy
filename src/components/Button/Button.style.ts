import {StyleSheet} from 'react-native';
export const style = StyleSheet.create({
  btn: {
    marginVertical: 10,
    paddingHorizontal: 16,
    borderRadius: 4,
    height: 48,
    alignItems: 'center',
    justifyContent: 'center',
  },
  btnText: {
    textAlign: 'center',
    textTransform: 'uppercase',
    fontSize: 13,
    fontWeight: '500',
  },
  fillbtnText: {
    color: '#FFFFFF',
  },
  btnPrimary: {
    backgroundColor: '#28AF51',
  },
  btnGhost: {
    backgroundColor: '#F8F8FA',
  },
  btndanger: {
    backgroundColor: '#E30E2C',
  },
  btnGhostText: {
    color: '#000',
  },
  OutlineBtnDanger: {
    borderColor: '#E30E2C',
  },
  disabled: {
    opacity: 0.5,
  },
});
