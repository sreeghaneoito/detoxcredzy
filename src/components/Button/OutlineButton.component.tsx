/* eslint-disable react-native/no-inline-styles */
import {ActivityIndicator, Text, TouchableOpacity, View} from 'react-native';
import React from 'react';
import {style} from './Button.style';
import {useTheme} from '../../hooks/useTheme';
import {ButtonProps} from '../../interfaces/component.interface';

/**
 * Outline Button
 * @param type: Button type. it should be primary, ghost, or danger
 * @param label: Button name
 * @param Customstyle: Custom style for button. it should be a style object
 * @param triggerclick: Function that trigger on button click
 * @param disabled: Is the button is disabled or not
 * @param testid: Test id for automation
 * @param loading: Is the button is loading or not
 * @returns JSX Component
 */
const OutlineButton = ({
  Customstyle,
  triggerclick,
  label,
  type,
  testid,
  loading,
}: ButtonProps) => {
  const {theme} = useTheme();
  try {
    return (
      <TouchableOpacity
        testID={`${testid}btn`}
        onPress={triggerclick}
        style={[
          style.btn,
          {
            borderWidth: 1,
            backgroundColor: 'transparent',
          },
          Customstyle,
          type === 'primary' && theme.outlineButton,
          type === 'danger' && style.OutlineBtnDanger,
        ]}>
        {loading ? (
          <ActivityIndicator />
        ) : (
          <Text style={[style.btnText, theme.onSurface]}>{label}</Text>
        )}
      </TouchableOpacity>
    );
  } catch (error) {
    console.log('Outline Button crash =>', error);
    return <View style={style.btn} />;
  }
};

export default OutlineButton;
