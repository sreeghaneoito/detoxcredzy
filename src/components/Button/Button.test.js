import React from 'react';
import renderer from 'react-test-renderer';
import Button from './Button.component';

test('renders correctly', () => {
  const button = renderer.create(<Button />).toJSON();
  expect(button).toMatchSnapshot();
});
