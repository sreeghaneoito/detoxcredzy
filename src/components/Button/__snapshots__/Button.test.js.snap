// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "borderRadius": 4,
      "height": 48,
      "justifyContent": "center",
      "marginVertical": 10,
      "opacity": 1,
      "paddingHorizontal": 16,
    }
  }
  testID="undefinedbutton"
>
  <Text
    collapsable={false}
    style={
      Object {
        "color": "#FFFFFF",
        "fontSize": 13,
        "fontWeight": "500",
        "textAlign": "center",
        "textTransform": "uppercase",
      }
    }
  />
</View>
`;
